### ISSUE TITLE
jsonnet vendoring

### ISSUE THREAD

**Issue by jacobstr** [google/jsonnet#521]
TLDR;  Is there an official stance/roadmap on vendoring/sharing jsonnet libraries?

We've recently eschewed using ksonnet because of some less-than-positive impressions with the generated API code. We've also got somewhat different requirements for how we lay our resources out than prescribed by the `ks` tooling.

Still, having a shared set of libraries around kubernetes concepts like deployments, service, etc is a good thing. To that end we've written our own set of helpers around kubernetes resources.

For example:

```
  /* deployment returns an apps/v1.Deployment resource.

    Params:
      meta: resource.metadata().
      selector: resource.selector().
      replicas: Integer, desired number of pods.
      template: resource.pod().
      strategy: Optional resource.updateStrategy().
  */
  deployment(meta, selector, replicas, template, strategy=null):: {
    apiVersion: 'apps/v1',
    kind: 'Deployment',
    metadata: meta,
    spec: {
      selector: selector,
      replicas: replicas,
      template: template,
      [if strategy != null then 'strategy']: strategy,
    },
  },
```

As much as possible, I'd like to encourage internal users to use this library rather than re-inventing their own. To that end, I'd like to be able to maintain our jsonnet helper libraries in a separate repository and have a familiar workflow around sharing this code. I.e. users can install libraries from known locations, we have some idea of provenance (what repo it came from, what ref it corresponds to), some hygiene around avoiding local modifications and forks (don't edit the contents of the vendor folder!).

It looks like the ksonnet folks have tackled this  e.g. with `ks pkg install incubator/redis@master`.

There's a nascent third-party project out there, https://github.com/jsonnet-bundler/jsonnet-bundler, that attempts to push this functionality down into the jsonnet realm. IMO this feels like a great place for it - not every json-authoring helper library has to be strictly kubernetes oriented. The project is quite alpha but it solves some of the problems around a jsonnet vendoring workflow.

Some low-fi solutions could be to use git submodules for example. We've had poor experiences with those: folks forget to init/update, folks tend to forget to push changes (the diffs are deceptive).

Cargo-culting the libraries leads to a diaspora of local modifications without history of the point in time where the copy/pasting occurred. You _can_ make local modifications to a vendor folder in golang, but it evokes the same "danger" sensation as black\yellow stripes on a wasp to do so.

**Comment by sbarzowski** [comment]
You pretty much covered it all 😄. 

The jsonnet-bundler project is almost-official. It was consulted with @sparkprime and broader community and it has everyone's blessing. So we are counting on jsonnet-bundler becoming the standard for jsonnet dependency management. However, as you mentioned, it's still pretty alpha.

Ksonnet and some other projects are going with custom solutions, specific to their case. 

And other than that, there are various low-tech solutions, not specific to Jsonnet in any way (submodules, copy in the repo, bash script which downloads the library, etc...).

(@sparkprime right?)

**Comment by brancz** [comment]
Specifically regarding ksonnet, we talked extensively with the ksonnet maintainers and they're all for adopting jsonnet-bundler, especially if we can provide it as a library as well.

That said, yes jsonnet-bundler is definitely very alpha, but a number of companies are already using it on a day to day basis. There are lots of unimplemented features, and we haven't tackled the distribution/registry problem that was covered in the original [design doc](https://docs.google.com/document/d/1czRScSvvOiAJaIjwf3CogOULgQxhY9MkiBKOQI1yR14/edit) at all.

That said I feel comfortable as we use jsonnet-bundler every day ourselves that over the upcoming months we'll eventually add all the features we have described in the doc, and then are in a position where we can promote jsonnet-bundler to be a more stable and officially endorsed solution.

**Comment by sbarzowski** [comment]
I'm going to close this issue. The current solution is to either use Jsonnet bundler or submodules or integrate Jsonnet with whatever custom mechanism you're using (if you have one). The language itself is unopinionated about the distribution of libraries.
